ipa-core/src/protocol/ipa_prf/shuffle/malicious.rs

```diff
@@ -354,12 +470,138 @@ mod tests {
             boolean_array::{BA112, BA144, BA20, BA32, BA64},
             Serializable,
         },
+        helpers::in_memory_config::{MaliciousHelper, MaliciousHelperContext},
         protocol::ipa_prf::shuffle::base::shuffle,
         secret_sharing::SharedValue,
         test_executor::run,
-        test_fixture::{Reconstruct, Runner, TestWorld},
+        test_fixture::{Reconstruct, Runner, TestWorld, TestWorldConfig},
     };
 
+    /// Test the hashing of `BA112` and tag equality.
+    #[test]
+    fn hash() {
+        run(|| async {
+            let world = TestWorld::default();
+
+            let mut rng = thread_rng();
+            let record = rng.gen::<BA112>();
+
+            let (keys, result) = world
+                .semi_honest(record, |ctx, record| async move {
+                    // compute amount of MAC keys
+                    let amount_of_keys: usize = (usize::try_from(BA112::BITS).unwrap() + 31) / 32;
+                    // // generate MAC keys
+                    let keys = (0..amount_of_keys)
+                        .map(|i| ctx.prss().generate_fields(RecordId::from(i)))
+                        .map(|(left, right)| AdditiveShare::new(left, right))
+                        .collect::<Vec<AdditiveShare<Gf32Bit>>>();
+
+                    // compute and append tags to rows
+                    let shares_and_tags: Vec<AdditiveShare<BA144>> = compute_and_add_tags(
+                        ctx.narrow(&OPRFShuffleStep::GenerateTags),
+                        &keys,
+                        iter::once(record),
+                    )
+                    .await
+                    .unwrap();
+
+                    (keys, shares_and_tags)
+                })
+                .await
+                .reconstruct();
+
+            let result_ba = BA112::deserialize_from_slice(&result[0].as_raw_slice()[0..14]);
+
+            assert_eq!(record, result_ba);
+
+            let tag = <BA112 as TryInto<Vec<Gf32Bit>>>::try_into(record)
+                .unwrap()
+                .iter()
+                .zip(keys)
+                .fold(Gf32Bit::ZERO, |acc, (entry, key)| acc + *entry * key);
+
+            let tag_mpc = <BA32 as TryInto<Vec<Gf32Bit>>>::try_into(BA32::deserialize_from_slice(
+                &result[0].as_raw_slice()[14..18],
+            ))
+            .unwrap();
+            assert_eq!(tag, tag_mpc[0]);
+        });
+    }
+
+    /// This test checks the correctness of the malicious shuffle.
+    /// It does not check the security against malicious behavior.
+    #[test]
+    fn check_shuffle_correctness() {
+        const RECORD_AMOUNT: usize = 10;
+        run(|| async {
+            let world = TestWorld::default();
+            let mut rng = thread_rng();
+            // using Gf32Bit here since it implements cmp such that vec can later be sorted
```

Comment:
You can sort any BA up to 128 bits with `records.sort_by_key(BA32::as_u128)`. This saves the effort of converting back and forth with `Gf32Bit`, and makes it pretty easy to add a correctness check to the `BA112` test. (Or perhaps to combine the tests into a generic test using `S: BooleanArray + U128Conversions`.)